src/clj/puppetlabs/services/jruby_pool_manager/impl/jruby_agents.clj

```diff
@@ -86,87 +87,102 @@
    pool :- jruby-schemas/pool-queue-type]
   (= expected-pool-size (count (.getRegisteredElements pool))))
 
-(schema/defn ^:always-validate
-  swap-and-drain-pool!
-  "Replace the current pool with a new pool and drain the old pool,
-  optionally refilling the new pool with fresh jrubies."
+(schema/defn collect-all-jrubies :- [JRubyInstance]
+  "Locks the pool and borrows all the instances"
+  [pool-state :- jruby-schemas/PoolState]
+  (let [pool-size (:size pool-state)
+        pool (:pool pool-state)
+        borrow-fn (fn [_] (jruby-internal/borrow-from-pool!*
+                           jruby-internal/borrow-without-timeout-fn
+                           pool))]
+    (.lock pool)
+    (try
+      (mapv borrow-fn (range pool-size))
+      (catch Exception e
+        (.clear pool)
+        (.insertPill pool (PoisonPill. e))
+        (throw (IllegalStateException.
+                "There was a problem borrowing a JRubyInstance from the pool."
+                e)))
+      (finally
+        (.unlock pool)))))
+
+(schema/defn cleanup-and-refill-pool
+  "Cleans up the given instances and optionally refills the pool with
+  new instances. Should only be called from the modify-instance-agent"
   [pool-context :- jruby-schemas/PoolContext
-   old-pool-state :- jruby-schemas/PoolState
-   new-pool-state :- jruby-schemas/PoolState
+   pool-state :- jruby-schemas/PoolState
+   old-instances :- [JRubyInstance]
    refill? :- schema/Bool]
-  (let [{:keys [config]} pool-context
-        pool-state-atom (jruby-internal/get-pool-state-container pool-context)
-        new-pool (:pool new-pool-state)
-        old-pool (:pool old-pool-state)
-        old-pool-size (:size old-pool-state)
+  (let [pool (:pool pool-state)
+        pool-size (:size pool-state)
+        new-instance-ids (map inc (range pool-size))
```

Comment:
@camlow325 ahh, ok. Yeah we should preserve that then. There's a `next-instance-id` fn that guarantees a unique new id based off the old id. We could use that if we don't want to use an atom.

Sounds like a good followup PR